### PR TITLE
Proposed solution to Issue #1

### PR DESCRIPTION
**Pull request by blaine-dodson**

### CODE REVIEW

File: src/utils.c [hunk 1]
```diff
@@ -49,14 +49,34 @@ int cat_file(char *path) {
...
-
-	tracef[0] = 0;
-	strcat(tracef, tracefs);
-	strcat(tracef, "/");
-	strcat(tracef, fn);
-
-	return cat_file(tracef);
+/*	char tracef[100];*/
+
+/*	tracef[0] = 0;*/
```
Comment (joelagnel) [inline src/utils.c]: why are the comments still here?

File: src/utils.c [hunk 2]
```diff
@@ -49,14 +49,34 @@ int cat_file(char *path) {
...
+	
+	len_tfs = strlen(tracefs);
+	len_fn  = strlen(fn);
+	
+	/* 1 for the '\0', and one for the '/' */
+	tracef = malloc(len_tfs+len_fn+2);
+	
+	/* I opted to use memcpy instead of strcat because I only wanted to
+	itterate each string once. */
+	memcpy(tracef, tracefs, len_tfs);
```
Comment (joelagnel) [inline src/utils.c]: strcat

File: src/utils.c [hunk 3]
```diff
@@ -49,14 +49,34 @@ int cat_file(char *path) {
...
+	memcpy(tracef+len_tfs+1, fn, len_fn);
+	
+	tracef[len_tfs] = '/';
+	tracef[len_tfs+len_fn+1] = '\0';
+	
+	return_val = cat_file(tracef);
```
Comment (joelagnel) [inline src/utils.c]: thanks for the effort, could you follow the style guidelines before posting:
https://www.kernel.org/doc/html/v4.10/process/coding-style.html

File: src/utils.c (hunk 2 above)
Comment (joelagnel) [reply src/utils.c]: The other way might be, just use strncat and then we don't need to malloc at all.

**Review by joelagnel — CHANGES_REQUESTED** [review]

File: src/bpfd.c [hunk 4]
```diff
@@ -65,25 +65,40 @@ int bpf_prog_load_handle(int type, char *name, char *bin_b64, int prog_len, char
 
 int get_trace_events(char *tracefs, char *category)
 {
-	char tracef[256];
-
+	char * tracef;
```
Comment (joelagnel) [inline src/bpfd.c]: char *tracef;

**Comment by blaine-dodson** [comment]
I don't understand how strncat could be a solution,
- the size parameter only applies to the source string
- I would still have to use strlen to keep a running tally of space remaining in the destination buffer
- on failure strncat would silently truncate

If we're committed to the fixed length buffers I could also just check total length first and fail before calling strcat.

I can push another commit when I get clarification. I'll also clean up the formatting as per your request.

**Comment by joelagnel** [comment]
Why?
`strncat` prevents overflow on the destination, by ensuring we don't cross its boundaries.
See: [https://linux.die.net/man/3/strncat](https://linux.die.net/man/3/strncat)
```
char tracef[256];
strncat(tracef, tracefs, 256)
strncat(tracef, '/', 256)
strncat(tracef, category, 256)
```
The only drawback is we'rd hardcoding the length, but atleast that's better than dynamic memory allocation failure for any reason IMO.

**Comment by blaine-dodson** [comment]
In principle I agree with you that strncat would be most useful if it protected the destination buffer from overflow. However, as I said previously, that is not how strncat works. Quoting from the [link](https://linux.die.net/man/3/strncat) you sent me:

"...the dest string must have enough space for the result. If dest is not large enough, program behavior is unpredictable; buffer overruns are a favorite avenue for attacking secure programs. "

strncat's size parameter limits the number of characters read from the **source** string, not the total number of characters written to the destination string. For example, the following code crashes because of stack smashing. Each string individually is shorter than the limit, but together they overflow the destination buffer.

#include <string.h>
#include <stdio.h>
#define BUF 20

int main(){
	char tracef[BUF];
	
	strncat(tracef, "FIRST STRING. ", BUF);
	strncat(tracef, "SECOND STRING.", BUF);
	puts(tracef);
	return 0;
}

I have reverted my two previous commits since they were unsatisfactory and I thought it better to start fresh--they weren't that great anyway. I then made a single commit that should correct issue #1 without using dynamic allocation by checking the total length before actually writing to the buffer.

Please let me know if I screwed anything else up.

**Comment by joelagnel** [comment]
Yes you screwed up, and you are reading the wrong documentation and wasting my time.
The documentation you quoted is for `strcat`. `strncat` on the other hand makes sure not to overflow the destination. The link I sent you is documentation for both `strcat` and `strncat`.
Also no need to revert commits that were never merged. You can just rebase your local branch and remove the commits you don't want. I am not going to merge your revert commits from your PR :)

**Comment by blaine-dodson** [comment]
Sorry. I'm done.

**Pull request closed without merging**